### ISSUE TITLE
Company tng map should address clash with templates map

### ISSUE THREAD

**Issue by rski** [company-mode/company-mode#731]
I've added (company-tng-configure-default) to my init like the docs in company-tng.el recommend, but the way TAB is bound, it clashes with TAB inside templates.

Consider this really contrived case:

![2017-10-19-175507_316x49_scrot](https://user-images.githubusercontent.com/2960312/31783590-aa91807c-b4f6-11e7-9a3a-793f095fff32.png)

Here, my instinctive reaction is to press tab to select fmt, but because the point is inside a template, it jumps to the next field. I've solved it for go by telling the backend not to insert a template, but the docs should probably be updated with a recommendation to work around this and tng-configure-default should do something about it too.

I would solve it by unbinding TAB from the template map, but what should forward-field be bound to instead?

**Comment by dgutov** [comment]
It's an old problem, but it's a lot more noticeable with tng, of course.

Yasnippet's keymap takes priority because it's set using an overlay. Last I checked, setting it this way doesn't work at the last buffer position, so that's why we still use the current option. Keeping both approaches and switching between them dynamically could work, though.

Maybe some rebinding would satisfy most people, though. Problem is, we'd have to do it for both `company-template` and `yasnippet`, which would be pretty invasive. Paging @nikital.

**Comment by nikital** [comment]
Hi, good morning :)

I'm not using snippets at the moment so I'm not sure what the desired default behavior should be...

1. TAB will select the candidate—`company-tng-configure-default` will unbind the "next template field" keybindings.
2. TAB will move to the next template field—`company-tng` will deactivate itself when a template is active.
3. TAB will do some sophisticated DWIM action? I can't think of a reasonable DWIM behavior.
4. `company-tng-configure-default` will disable templates entirely by default to avoid out-of-the-box confusion for new users.

I lean towards option 1 and asking the user to bind some other key for "next field". In Vim I bind C-j for snippet expansion and next template field, while keeping TAB exclusively for completions.

Which option sounds the most reasonable to you? Do you see other options?

**Comment by rski** [comment]
Heya @nikital 

I'm not using the yas backend either, so it's good that @dgutov brought it up because it would've completely slipped my mind.

Putting aside yas for a second, I think that (1) is the most reasonably consistent behaviour where you don't have to do the mental shift from "TAB completes" to "i'm in a template now, TAB does another thing".

As for inside yas, probably a best effort approach is good enough. Just let the tng users know that TAB clashes with yas snippet and recommend they bind it to a different keybinding* in the tng docstring. I agree that having company rebind yas' TAB would be too invasive.

* C-j might not be the best recommendation because IIRC some major modes use this for things. Maybe have the docstring recommend they unbind M-n M-p from company candidate selection and use that for yas?

**Comment by alexmurray** [comment]
@nikital Can you advise how to easily unbind tab from yasnippet and then assign a new keybinding for yas-next-field etc so this doesn't conflict with company-tng?

**Comment by alexmurray** [comment]
@nikital - ignore the above - I figured out a way which works for me - https://github.com/alexmurray/dot_emacs.d/commit/5db544b43f1cc821443bf8bc57a60f90e3f5d72e

**Comment by nikital** [comment]
@rski I don't use company-template, so I took a clean Emacs configuration and studied the interaction to get a general idea of how it looks like. Apart from the TAB issue you mentioned, I see another problem:
```c
int hello(int a, int b) { return a + b; }
int main ()
{
  hel|//point here, completion tooltip for hello visible
}
```
When we hit TAB, `hello` will be inserted into the buffer. Next we would type "(" to write a function invocation, at which point the template will fire, inserting another "(":
```c
int hello(int a, int b) { return a + b; }
int main ()
{
  hello((|, int b)
}
```
Did you notice it when you rebound TAB? If so, what did you do about that?

**Comment by dgutov** [comment]
@nikital That's the default behavior with company-tng now, isn't it? One probably gets used to it, and starts typing the first argument in the method call, instead of the opening paren.

If that's not good enough, we can dd a global option that enables or disables snippet expansion (then company-tng would set it to nil; in that case we'll probably need a new command that does perform expansion).

Ideally, we'd replace most `post-completion` callbacks with a standardized `snippet` callback (we'll either have to depend on yasnippet, or reimplement a snippet templating language in company-template). Then we even could render snippet previews.

**Comment by nikital** [comment]
> That's the default behavior with company-tng now, isn't it? One probably gets used to it, and starts typing the first argument in the method call, instead of the opening paren.

I think this is still not a solution because you don't always want to expand a template when you complete a function by name. For example, when you pass a function as a callback to another function a forced expansion after a completion would be annoying.

> If that's not good enough, we can dd a global option that enables or disables snippet expansion (then company-tng would set it to nil; in that case we'll probably need a new command that does perform expansion).

This can work as a patch for now

> Ideally, we'd replace most post-completion callbacks with a standardized snippet callback (we'll either have to depend on yasnippet, or reimplement a snippet templating language in company-template). Then we even could render snippet previews.

I'm not a user of snippets/templates so I don't know what it really means. I plan to look into integrating snippets into my workflow some time in the future and by then I will probably have an opinion about integrating it properly with company-tng. For now we can disable it if it causes user frustration.

BTW, do I understand correctly that company-template is another implementation of snippets in Emacs? Does it provide similar functionality to Yasnippets or are they different?

**Comment by dgutov** [comment]
> Does it provide similar functionality to Yasnippets or are they different?

It's a simple subset. It gives you a programmatic option to create input fields inside a buffer, but it doesn't implement any templating language (which would be required for this feature).

**Comment by nikital** [comment]
OK, I started looking into templates and snippets in order to solve it properly, feel free to assign the issue to me or label it or whatever you do to mark it in progress :)

**Comment by dgutov** [comment]
Thank you. I think your comment should suffice.

To be clear, you don't really have to do it, though. It seems like it's going to be a fair amount of work.

**Comment by nikital** [comment]
> To be clear, you don't really have to do it, though. It seems like it's going to be a fair amount of work.

Sounds like an opportunity to learn some Elisp so I'm fine with that :cake: 

I'm studying `company-template.el` and `yasnippet.el` and considering different ways forward. One of the things I'm considering is removing most of the template code from Company and pulling Yasnippet as a dependency for the rendering of the fields. I'm not sure what was the historical context that led to `company-template` (Was it created to avoid external dependencies? To make the code lighter? Yas wasn't around?) so I wonder if you'd be open for such a move or would you prefer to keep the templating code inside Company?

Also, **to make sure that I'm solving the actual problem I'll write here I think the problem is**:
Because company-tng binds TAB, it conflicts with field navigation in arguments fields from `company-template` and snippet fields from Yasnippet. Rebinding TAB requires messing with both libraries. We want to streamline the rebinding experience by uniting the two somehow and provide sane defaults for company-tng.

Are there more problems related to this that I didn't mention?

**Comment by dgutov** [comment]
> Was it created to avoid external dependencies? To make the code lighter? Yas wasn't around?

All three, possibly.

> So I wonder if you'd be open for such a move or would you prefer to keep the templating code inside Company?

I haven't reached a decision yet. Still, that can only affect the timeline of having your efforts merged, but not their general usefulness.

> Are there more problems related to this that I didn't mention?

I wasn't thinking about the rebinding issue at all (you can solve for both templating packages relatively easily, I think, as soon as you find a solution for one of them).

The problem I was thinking of is, to quote yourself:

> When we hit TAB, hello will be inserted into the buffer. Next we would type "(" to write a function invocation, at which point the template will fire, inserting another "(":

...and the way to solve it, I think, is to either disable snippet expansion when tng is used (to be frank, you could probably disable `post-completion` to almost the same effect), or show inline how each snippet will be expanded, so the user is not trying to type what's already going to be there.

**Comment by nikital** [comment]
1. I agree that disabling `post-completion` is a good first step, even for non-snippet backends.
2. What is the idea behind `pre-completion`? It was introduced with `post-completion` but was never used or documented.
3. What is the purpose of `company-completion-finished-hook` and `company-callback`? What do people usually use it for? (Because I'm wondering whether TNG should call the `finished` hook or the `cancelled` hook when completing)

**Comment by dgutov** [comment]
> What is the idea behind pre-completion? It was introduced with post-completion but was never used or documented

I don't know. In all likelihood, nobody uses it, and it can be deleted.

> company-completion-finished-hook

See https://github.com/company-mode/company-statistics/blob/master/company-statistics.el as one usage.

> and company-callback?

`company-begin-with` uses it. Which is a kind of alternative for `completing-read`. Depending on whether somebody, somewhere, uses it, we might want to deleted them both.

Maybe @nschum has something to add.

**Comment by dgutov** [comment]
The problem in the original description should be fixed in version 0.9.11.
